nautobot/core/ui/object_detail.py

```diff
@@ -0,0 +1,598 @@
+#      Panel ("Contact Associations")
+#  Tab ("Dynamic Groups")
+#    LayoutFullWidth
+#      Panel ("Dynamic Groups")
+#  Tab ("Object Metadata")
+#    LayoutFullWidth
+#      Panel ("Object Metadata")
+
+# ObjectDetailContent
+#  Tab ("IP Address")
+#    LayoutLeftHalf
+#      Panel ("IP Address")
+#      ...
+#    LayoutRightHalf
+#      Panel ("Operational Details")
+#      Panel ("Parent Prefixes")
+#    LayoutFullWidth
+#      Panel ("Related IP Addresses")
+#  ...
+
+# ObjectDetailContent
+#  Tab ("Location")
+#    LayoutLeftHalf
+#      Panel ("Location")
+#      Panel ("Geographical Info")
+#      ...
+#    LayoutRightHalf
+#      Panel ("Stats")
+#      Panel ("Rack Groups")
+#      Panel ("Images")
+#    LayoutFullWidth
+#      Panel ("Children")
+#  ...
+
+# ObjectDetailContent
+#  Tab ("Tenant")
+#    LayoutLeftHalf
+#      Panel ("Tenant")
+#      ...
+#    LayoutRightHalf
+#      Panel ("Stats")
+#  ...
+
+
+# A default detail view consists of:
+# ObjectDetailContent:
+#  Tab (model._meta.verbose_name|bettertitle)
+#    LayoutLeftHalf
+#      ObjectDetailsPanel(fields="__all__")
+#      (additional panels provided by Extras and Apps)
+#      ...
+#    LayoutRightHalf (empty, extensible by Apps)
+#    LayoutFullWidth (empty, extensible by Apps)
+#  (additional tabs provided by Core, Extras, and Apps)
+
+
+class ObjectDetailContent:
+    """
+    Base class for UI framework definition of the contents of an Object Detail (Object Retrieve) page.
+
+    This currently defines the tabs and their contents, but currently does NOT define the page title, breadcrumbs, etc.
+    """
+
+    def __init__(self, tabs=None, detail_fields=None, field_transforms=None):
+        """
+        Create an ObjectDetailContent.
+
+        Args:
+            tabs (list): Optional list of `Tab` instances; if unset, the default `ObjectDetailMainTab` will be used.
+                         Standard extras Tabs (advanced, contacts, dynamic-groups, metadata, etc.) do not need to be
+                         specified as they will be automatically included regardless.
+            detail_fields (dict): Optional dict of {panel_name: [list of object field names], ...}.
+                                  This addresses the common case where a detail view doesn't care to define custom tabs,
+                                  layouts, or panels, but wants some control over how the default `ObjectDetailMainTab`
+                                  presents the object's fields in an ObjectFieldsPanel(s).
+            field_transforms (dict): Optional dict of {field_name: [list of transform functions], ...}.
+                                     This can be used to provide custom rendering of given field(s) within the automatic
+                                     `ObjectDetailMainTab` class, for example `{"commit_rate": [humanize_speed]}`.
+        """
+        if tabs is None:
+            tabs = [ObjectDetailMainTab()]
+            if detail_fields is not None:
```

Comment:
I wonder if we should enhance this a little. For example, I can imagine it might be easy to miss certain fields from the model if the user has to declare it manually every time. We should probably compare the declared `detail_fields` to maybe the fields on the serializer and give off a warning in the template `messages.warning`?  But again, in the legacy template, it is also dependent on the user to declare fields in the template. So Im not sure.

Reply:
Yeah this is a tricky one to be sure. In the "default" case we'd just use `detail_fields="__all__"` but I'm not sure how many times that will actually be fit for purpose.

From just attempting the "Circuit" detail view, the `Comments` field is a separate `ObjectFieldsPanel` than the rest of the base fields, and the `circuit_termination_a` and `circuit_termination_z` fields are going to need a custom panel class of their own rather than being presented in an `ObjectFieldsPanel` at all. All of that is intentional and shouldn't result in a warning log that is noise and that the end user can't do anything about in any case. So I'm not sure how we'd automatically detect or warn developers if fields are inadvertently (as opposed to intentionally) omitted...

Hmm... maybe we could/should allow inverting this logic (similar to serializers) and have an `ObjectFieldsPanel` support an explicit `exclude` list of fields? Then we could warn if the ObjectDetailContent as a whole doesn't contain at least one `ObjectFieldsPanel` with `fields=__all__, [exclude=...]` as its attributes?

Definitely bears some further thought.

Reply:
Added `exclude_fields` option in 7a794fdbc7785f78e71869d3173269374a473c93. We can refine this in a future PR.